debezium-core/src/main/java/io/debezium/relational/HistorizedRelationalSnapshotChangeEventSource.java

```diff
@@ -171,6 +223,111 @@ private boolean createSchemaChangeEventsForTables(ChangeEventSourceContext sourc
      */
     protected abstract SchemaChangeEvent getCreateTableEvent(SnapshotContext snapshotContext, Table table) throws Exception;
 
+    private boolean createDataEvents(ChangeEventSourceContext sourceContext, SnapshotContext snapshotContext) throws InterruptedException{
+        SnapshotReceiver snapshotReceiver = dispatcher.getSnapshotChangeEventReceiver();
+        snapshotContext.offset.preSnapshotStart();
+
+        for (TableId tableId : snapshotContext.capturedTables) {
+            if (!sourceContext.isRunning()) {
+                return false;
+            }
+
+            LOGGER.debug("Scanning table {}", tableId);
+
+            if (!createDataEventsForTable(snapshotContext, snapshotReceiver, snapshotContext.tables.forTable(tableId))) {
+                return false;
+            }
+        }
+
+        snapshotContext.offset.preSnapshotCompletion();
+        snapshotReceiver.completeSnapshot();
+        snapshotContext.offset.postSnapshotCompletion();
+
+        return true;
+    }
+
+    /**
+     * Dispatches the data change events for the records of a single table.
+     */
+    private boolean createDataEventsForTable(SnapshotContext snapshotContext, SnapshotReceiver snapshotReceiver,
+            Table table) throws InterruptedException {
+
+        long exportStart = clock.currentTimeInMillis();
+        LOGGER.info("\t Exporting data from table '{}'", table.id());
+
+        final String selectStatement = getSnapshotSelect(snapshotContext, table.id());
+        LOGGER.info("\t For table '{}' using select statement: '{}'", table.id(), selectStatement);
+
+        try (Statement statement = readTableStatement();
+                ResultSet rs = statement.executeQuery(selectStatement)) {
+
+            Column[] columns = getColumnsForResultSet(table, rs);
+            final int numColumns = table.columns().size();
+            int rows = 0;
+
+            while (rs.next()) {
+                rows++;
+                final Object[] row = new Object[numColumns];
+                for (int i = 0; i < numColumns; i++) {
+                    row[i] = getColumnValue(rs, i + 1, columns[i]);
+                }
+
+                if (rows % 10_000 == 0) {
```

Comment:
Should we do it based on thw row count or based on time, like every 2 seconds?

Reply:
That's copied from the MySQL `SnapshotReader`. No strong preference, I can change it, if you like.